edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/AuthAgentRequestHandler.cs

```diff
@@ -0,0 +1,377 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Net;
+    using System.Security.Cryptography.X509Certificates;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using Microsoft.AspNetCore;
+    using Microsoft.AspNetCore.Builder;
+    using Microsoft.AspNetCore.Hosting;
+    using Microsoft.AspNetCore.Http;
+
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Util;
+
+    using Microsoft.Extensions.DependencyInjection.Extensions;
+    using Microsoft.Extensions.Logging;
+    using Microsoft.Net.Http.Headers;
+
+    using Newtonsoft.Json;
+
+    public class AuthAgentRequestHandler
+    {
+        // The maximum allowed request size. The biggest item in a request is the certificate chain,
+        // 32k for the certificates and other data should be more than enough. Bigger requests are
+        // refused.
+        const int MaxInBufferSize = 32 * 1024;
+        const string Post = "POST";
+
+        static readonly TimeSpan requestTimeout = TimeSpan.FromSeconds(5);
+
+        readonly IAuthenticator authenticator;
+        readonly IUsernameParser usernameParser;
+        readonly IClientCredentialsFactory clientCredentialsFactory;
+        readonly AuthAgentProtocolHeadConfig config;
+
+        public AuthAgentRequestHandler(IAuthenticator authenticator, IUsernameParser usernameParser, IClientCredentialsFactory clientCredentialsFactory, AuthAgentProtocolHeadConfig config)
+        {
+            this.authenticator = Preconditions.CheckNotNull(authenticator, nameof(authenticator));
+            this.usernameParser = Preconditions.CheckNotNull(usernameParser, nameof(usernameParser));
+            this.clientCredentialsFactory = Preconditions.CheckNotNull(clientCredentialsFactory, nameof(clientCredentialsFactory));
+            this.config = Preconditions.CheckNotNull(config, nameof(config));
+        }
+
+        public static IWebHost CreateWebHostBuilder(
```

Comment:
Does this work? 2 WebHosts in the same project?

Reply:
Yes, they are independent. I didn't want to merry it with the other one because that has its own middlewares/setup, also this is a separate component and felt cleaner having its own host (also was easier to implement and will be easier to maintain them if they evolve)